cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -698,18 +783,30 @@ Result<GcsOptions> GcsOptions::FromUri(const arrow::internal::Uri& uri,
     options_map.emplace(kv.first, kv.second);
   }
 
-  if (!uri.password().empty() || !uri.username().empty()) {
-    return Status::Invalid("GCS does not accept username or password.");
+  const std::string& username = uri.username();
+  bool anonymous = username == "anonymous";
+  if (!username.empty() && !anonymous) {
+    return Status::Invalid("GCS URIs do not accept username except \"anonymous\".");
+  }
+  if (!uri.password().empty()) {
+    return Status::Invalid("GCS URIs do not accept password.");
   }
+  auto options = anonymous ? GcsOptions::Anonymous() : GcsOptions::Defaults();
 
-  auto options = GcsOptions::Defaults();
   for (const auto& kv : options_map) {
     if (kv.first == "location") {
       options.default_bucket_location = kv.second;
     } else if (kv.first == "scheme") {
       options.scheme = kv.second;
     } else if (kv.first == "endpoint_override") {
       options.endpoint_override = kv.second;
+    } else if (kv.first == "retry_limit_seconds") {
+      int parsed_seconds = atoi(kv.second.c_str());
+      if (parsed_seconds <= 0) {
+        return Status::Invalid(
+            "retry_limit_seconds  must be a positive integer.  found '", kv.second, "'");
```

Comment:
Nit: error formatting
```suggestion
            "retry_limit_seconds must be a positive integer, got '", kv.second, "'");
```